src/module_wrap.cc

```diff
@@ -98,99 +98,87 @@ ModuleWrap* ModuleWrap::GetFromID(Environment* env, uint32_t id) {
+    context = sandbox->context();
+  }
+
   Local<Integer> line_offset;
   Local<Integer> column_offset;
 
-  if (argc == 5) {
-    // new ModuleWrap(source, url, context?, lineOffset, columnOffset)
-    if (args[2]->IsUndefined()) {
-      context = that->CreationContext();
-    } else {
-      CHECK(args[2]->IsObject());
-      ContextifyContext* sandbox =
-          ContextifyContext::ContextFromContextifiedSandbox(
-              env, args[2].As<Object>());
-      CHECK_NOT_NULL(sandbox);
-      context = sandbox->context();
-    }
-
+  bool synthetic = args[2]->IsArray();
+  if (synthetic) {
+    // new ModuleWrap(url, context, exportNames, syntheticExecutionFunction)
+    CHECK(args[3]->IsFunction());
+  } else {
+    // new ModuleWrap(url, context, source, lineOffset, columOffset)
+    CHECK(args[2]->IsString());
     CHECK(args[3]->IsNumber());
     line_offset = args[3].As<Integer>();
-
     CHECK(args[4]->IsNumber());
     column_offset = args[4].As<Integer>();
-  } else {
-    // new ModuleWrap(source, url)
-    context = that->CreationContext();
-    line_offset = Integer::New(isolate, 0);
-    column_offset = Integer::New(isolate, 0);
   }
 
-  Local<String> url;
-  Local<Module> module;
-
   Local<PrimitiveArray> host_defined_options =
       PrimitiveArray::New(isolate, HostDefinedOptions::kLength);
   host_defined_options->Set(isolate, HostDefinedOptions::kType,
                             Number::New(isolate, ScriptType::kModule));
 
-  // new ModuleWrap(syntheticExecutionFunction, export_names, url)
-  bool synthetic = args[0]->IsFunction();
-  if (synthetic) {
-    CHECK(args[1]->IsArray());
-    Local<Array> export_names_arr = args[1].As<Array>();
-
-    uint32_t len = export_names_arr->Length();
-    std::vector<Local<String>> export_names(len);
-    for (uint32_t i = 0; i < len; i++) {
-      Local<Value> export_name_val =
-          export_names_arr->Get(context, i).ToLocalChecked();
-      CHECK(export_name_val->IsString());
-      export_names[i] = export_name_val.As<String>();
-    }
+  ShouldNotAbortOnUncaughtScope no_abort_scope(env);
+  TryCatchScope try_catch(env);
 
-    CHECK(args[2]->IsString());
-    url = args[2].As<String>();
+  Local<Module> module;
 
-    module = Module::CreateSyntheticModule(isolate, url, export_names,
+  {
+    ScriptOrigin origin(url,
+                        line_offset,                      // line offset
+                        column_offset,                    // column offset
+                        True(isolate),                    // is cross origin
+                        Local<Integer>(),                 // script id
+                        Local<Value>(),                   // source map URL
+                        False(isolate),                   // is opaque (?)
+                        False(isolate),                   // is WASM
+                        True(isolate),                    // is ES Module
+                        host_defined_options);
```

Comment:
Do we have to construct this `ScriptOrigin` for the synthetic case? Seems unnecessary...

Reply:
You're right, when I wrote this I had forgotten that synthetic modules don't take origins as an argument.